File: gpt-4-turbo/stage2/serial_construct_num_workers_clause.cpp
#include <iostream>
#include <stdlib.h>
#include "acc_testsuite.h"

#ifndef T1
//T1:parallel, construct, num_workers clause, V:2.7-3.3
int test1(){
    int err = 0;
    srand(SEED);

    const int size = 1000;
    int *a = new int[size];
    int *b = new int[size];
    int expectedSum = 0, actualSum = 0;

    for (int i = 0; i < size; ++i) {
        a[i] = rand() % 100;
        expectedSum += a[i];
    }

    #pragma acc parallel num_workers(4) copyout(b[0:size])
    {
        #pragma acc loop
        for(int i = 0; i < size; ++i) {
            b[i] = a[i];
        }
    }

    for (int i = 0; i < size; ++i) {
        actualSum += b[i];
    }

    if (expectedSum != actualSum) {
        err += 1;
    }

    delete[] a;
    delete[] b;

    return err;
}
#endif

int main(){
    int failcode = 0;
    int failed;
#ifndef T1
    failed = 0;
    for (int x = 0; x < NUM_TEST_CALLS; ++x){
        failed = failed + test1();
    }
    if (failed != 0){
        failcode = failcode + (1 << 0);
    }
#endif
    return failcode;
}

File: gpt-4-turbo/stage2/serial_construct_num_workers_clause.f90
!T1:serial construct num_workers clause,V:2.7-2.3
      LOGICAL FUNCTION test1()
        USE OPENACC
        IMPLICIT NONE
        INCLUDE "acc_testsuite.Fh"
        
        INTEGER :: errors = 0
        INTEGER :: x = 0

        !$acc serial num_workers(1)
        x = x + 1
        !$acc end serial

        ! The actual number of workers is transparent in serial context.
        ! Verification is by ensuring no runtime errors and correct execution.
        IF (x /= 1) THEN 
          errors = errors + 1
        END IF

        IF (errors .eq. 0) THEN
          test1 = .FALSE.
        ELSE
          test1 = .TRUE.
        END IF
      END FUNCTION test1

      PROGRAM main
        USE OPENACC
        IMPLICIT NONE
        INTEGER :: failcode = 0, testrun
        LOGICAL :: failed
        INCLUDE "acc_testsuite.Fh"
        LOGICAL :: test1
        failed = .FALSE.
        
        DO testrun = 1, NUM_TEST_CALLS
          failed = failed .or. test1()
        END DO
        IF (failed) THEN
          failcode = failcode + 2 ** 0
          failed = .FALSE.
        END IF
        CALL EXIT (failcode)
      END PROGRAM main